### DISCUSSION TITLE
`on_dag_run_running` (via @hookimpl) and database access.

### DISCUSSION THREAD

**Discussion by Blizzke** [apache/airflow#40296] [Q&A]
I've been spending quite some time on this and things do not seem logical.

When queuing a DAGRun and having a "global" `on_dag_run_running`-hook, any database related thing kills the actual state changes in the  parent session, so the run remains forever in the queued state.

I've been testing this pretty thoroughly and perhaps I'm overlooking something but I cannot find it.

The run begins its (relevant) part [in this piece of code](https://github.com/apache/airflow/blob/main/airflow/jobs/scheduler_job_runner.py#L1065) in `_do_scheduling` of the `scheduler_job_runner.py`:

```python
with prohibit_commit(session) as guard:
  ...
  self._start_queued_dagruns(session)
  guard.commit()
  ...
```

`_start_queued_dagruns` eventually calls `dag_run.notify_dagrun_state_changed()`, which ends up calling the actual hook.

I've added (between the call to `_start_queued_dagruns` and `guard.commit()` the following snippet:

```python
if session.dirty or session.new or session.deleted:
  self.log.error("There are pending changes in the session")
else:
  self.log.error("There are NO pending changes in the session")
```

Whatever I put in the hook that does not involve the database (for example a `sleep(5)`), after that I always get `"There are pending changes in the session"` and things continue as they should.

If I only create a local session and make sure that is closed, without doing any interaction (like so):
```python
@hookimpl
def on_dag_run_running(dag_run):
  from airflow.settings import Session
  s = Session()
  sleep(5)
  s.close()
```

The output string after that is `"There are NO pending changes in the session"`. So the act of creating a session instance within the hook seems to clear changed state (the DAGrun status change from `queued` to `running`) from the `CommitProhibitorGuard(session)`.

Note: I do not want to commit anything, I just need to read some of the extra DAGRun attributes, which are not loaded in the instance you receive.

I've found 2 ways to bypass this, but I would still like to know if I'm doing something wrong.

For those who need it: If you move the code from the hook to either a new thread (`threading.Thread` or `multiprocessing.Process`) things work. Alternatively, creating a completely separate `sessionmaker` also works.

To be honest, I don't like either of those solutions, but a separate thread/process is the least invasive.
Any ideas, suggestions?

Regarding that `sessionmaker` for those who want it:
```python
from sqlalchemy.orm import sessionmaker
from airflow.settings import engine

Session = sessionmaker(bind=engine)
session = Session()
```

**Comment by potiuk** [comment]
It's a dangerous thing to rely on database structure of Airlfow. And you should avoid doing it. This is very explicitly specified in https://airflow.apache.org/docs/apache-airflow/stable/public-airflow-interface.html#what-is-not-part-of-the-public-interface-of-apache-airflow -> the structure, types of fields, model names etc. WILL continue to change.

If you need to get additional information from airflow you should use one of the two things:

1) Use REST API https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html to retrieve the information you need. Yes it adds overhead, but the REST API is designed specifically to shield you from the implementation details of Airflow DB and airflow internals

2) prepare the data elsewhere - in another storage - and use it from there.

In Airlfow 3 Accessing Airflow DB from your code will be completely blocked. You won't be able to interact with the DB at all. You will not be able to create a session, you will not be able to use any of Airlfow Models. You will not have direct DB access. Full Stop. Non negotiable. https://cwiki.apache.org/confluence/display/AIRFLOW/%5BWIP%5D+AIP-72+Task+Execution+Interface+aka+Task+SDK.

So whatever you do now and want to be sustainable in the future, DON't create session, DON't use Airflow models. If things there work for you now, they work incidentally and they will stop working in the future.,

**Comment by Blizzke** [comment]
I do not rely on "the database structure", I start a new session so I can merge the `dag_run` into it which allows me to access the lazily loaded attributes that I need for my handling. We use the tasks and the xcom data for a lot of extra functionality.

And if blocking the DB access is the plan for Airflow 3, then we'll stick with 2.x until we are certain we can do everything in there that we are doing now.

You make a point of replying something along these lines on pretty much every of my my interactions here, but what's the use of being able to write plugins if you'll have to fetch all your extra data via the API? The only thing that I know for sure is that - with what the API AF is offering now - we don't come close to the functionality we use day to day, or definitely not without a huge amount of extra effort.

**Comment by potiuk** [comment]
> You make a point of replying something along these lines on pretty much every of my my interactions here, but what's the use of being able to write plugins if you'll have to fetch all your extra data via the API? The only thing that I know for sure is that - with what the API AF is offering now - we don't come close to the functionality we use day to day, or definitely not without a huge amount of extra effort.

Well. You can argue with me, but you should not argue with:

> [Database structure](https://airflow.apache.org/docs/apache-airflow/stable/database-erd-ref) is considered to be an internal implementation detail and you should not assume the structure is going to be maintained in a backwards-compatible way.

And yes it is more work for you of course. 

And yes I mention it every time anyone mentions that they are using the DB because when We released Airflow 2 we made it a decision that DB is not going to be supported "interface" for Airflow users and we introduced REST API for that. And yes, you might contest it and stick with Airlfow 2 if you wish. 

Also one more thign - my comments are not for you. Those comments are directed to any other user of Airlfow who finds this issue and will get impression that accessing Airflow Metadata DB by user code is fine in Airlfow. It's not. We (as community) made a deliberate decision and voted on it 4 yeaars ago when we released Airflow 2 and released REST API. So I want to make it crystal clear to anyone reading that issue now and in the future, that this is bad approach. 

That's why I comment like that on every single issue that could give impression that accessing Airlfow Metadata DB is ok. It's not and it's been decided by the community years ago.

**Comment by potiuk** [comment]
> You make a point of replying something along these lines on pretty much every of my my interactions here, but what's the use of being able to write plugins if you'll have to fetch all your extra data via the API? The only thing that I know for sure is that - with what the API AF is offering now - we don't come close to the functionality we use day to day, or definitely not without a huge amount of extra effort.

And thing about it here - over the 4 years the API has evolved and became way more powerful. If you miss anything there, creating an issue describing your need and rationale, or even better - creating a PR implementing it, is the right way to go. You had 4 years for that. Finally we are getting to the point where Airflow 3 is going to enforce it.

If you say 4 years is not enough, and you have not seen it coming and stayed with the old approach rather than helping to make REST API better for what you need, I'd say, you had a lot of chances that you have not used.

**Reply by Blizzke** [reply]
> And thing about it here - over the 4 years the API has evolved and became way more powerful. If you miss anything there, creating an issue describing your need and rationale, or even better - creating a PR implementing it, is the right way to go. You had 4 years for that. Finally we are getting to the point where Airflow 3 is going to enforce it.
> 
> If you say 4 years is not enough, and you have not seen it coming and stayed with the old approach rather than helping to make REST API better for what you need, I'd say, you had a lot of chances that you have not used.

We've tried, for example for the Dataset API endpoints - something that is of vital importance for how we see our way of working in the future - and basically that was shut down. Also the question of `extra` support per dataset event (instead of only once for the initial dataset record) was shut down (both by you iirc). 

So forgive me if that doesn't inspire me to continue along that path. We, as a team work, together with a number of other teams and we try to feel our way forward into a cooperation that works for everyone, of wich the biggest part of people has nothing to do with the "airflow stack" if I might say.. 

Our experience regarding the support here is "tunnel vision". Which is why I tried a discussion instead of a ticket this time, only to end up at the same spot. It's your right I guess, as you support the product, and I'm just the one whining, but that doesn't mean that the way things are decided here work for our company and the intra-team cooperation.

**Reply by potiuk** [reply]
Again. You make arguments 'ad personna' and I am telling you that you are arguing with the whole community that took different decisions.

While you are criticizing people (me) - I am criticizing the approach and design decision . That is quite bad approach for discussion that does not move the needle.

I suggest you take a look at our mukti team approach AIP-67 that should address your need way better than the API you proposed:
 https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-67+Multi-team+deployment+of+Airflow+components

And contributing to it might be a good idea while it is not yet finally approved (by the community)

Yes it is different than what you came with and yes your proposal (I did not even remember who proposed it - because I do not look at the authorship but about merit of the proposal) was not in line with our (community) vision where we are heading (a d some foundations for it were laid 4 years ago).

The thing is that adding a bandaid that we would have to support for many years when we are just about to define Airflow 3 vision is wrong.  We want to get people to contribute - we have Airflow Improvement Proposals to discuss, devlist discussions and weekly dev calls - and you are absolutely welcome to come there and make your proposals. They might be rejected (be prepared for defending them) or accepted - it's based on the merit of those proposals, not the people who raise them.

And I strongly encourage you to take part in this discussions if you want to shape future Airflow. But accessing the DB of Airflow directly by users have been discussed and rejected 4 years ago and we are only doubling down on it (as a community). So I suggest you at least pay atention to this discussions - but contributing is better.

**Reply by Blizzke** [reply]
The AIP you keep quoting (AIP 72) isn't even relevant in this discussion.
For one, it says nothing about plugins and/or hooks. As far as I can see is only related to workers and DAGs. This discussion/question has zero to do with that. 
Secondly: whatever the plans for 3.x might be, we are currently on 2.x, it's still being supported, so even if this question was about something that falls within that AIP, the question is valid at this moment and for this version.

That being said, I wanted to discuss about very peculiar SQLAlchemy behavior which might - or might not - be related on how airflow uses it. Perhaps it's even a bug that happens to show its ugly head within the lifecycle hook part of a plugin.
So basically, compare experiences here before taking this over to the SQLAlchemy community (if it turns out to be a bug). 
What I did not want was to end up in yet another discussion about what the future might/will bring for Airflow.

--- off topic ---

As for your other mention, the "multi team" -one, off topic of course but I did bring it up first: same story. 
We're a large company that works with 100s of gigabytes / terrabytes of data, from which the most part is still processed the "old fashioned way". Most teams are (unfortunately) isolated units that do their thing with their data and they almost all rely on Jenkins for the automation around it. Sending around "data events" was - for us - the easiest an least invasive way of distributing knowledge about updated data. 
So the fact that AF started supporting datasets for scheduling (and then the submitted PR for an API - and the commenting on it)  actually made us refactor things, as it seemed that this was a route that airflow as going to follow as well. But then the API PR was shut down without an alternative. 

Since we've already decided to go that way, and a giant company is slow moving/intert, we have to keep the course.
Also, that AIP assumes that everyone is using Airflow in some capacity, which is most definitely not the case here. 
As I already said: For me this seems a little myopic towards the mentality that all companies use Airflow for their data in some way. But again, if the community decided to take that path, so be it, it's just not one we can follow. And that is not because we don't want to, it's simply impossible. Before datasets we were being forced to use the API to trigger a specific DAG, which creates narrow dependencies between teams that we do not want.

But why am I even doing the effort to repeat myself.

**Reply by potiuk** [reply]
> Secondly: whatever the plans for 3.x might be, we are currently on 2.x, it's still being supported, so even if this question was about something that falls within that AIP, the question is valid at this moment and for this version.

Using DB structure has been discouraged in airflow 2 4 years ago. Reminding about it continuously is valid at this moment and for this version. Yes maybe there is a peculiar behaviour of sqlalchemy in this case but this is PRECISELY why we warned our users to not to try to use DB directly in their user code, because otherwise they - on their own will have to fight with those peculiarities. 

FYI: We have seen similar issues when we implemented listeners intitially as sqlalchemy hooks for openlineage and we did everything to move away from using DB access from within hook implementations, precisely because it caused similar problems. You might not have followed it before but you can take a look here https://github.com/apache/airflow/pull/29289

And since we learned that this is causing problems, we continue advising our users to not use sqlalchemy but switch to other means. If REST API is not enough - you probably missed my second suggestion, rather than accessing the DB directly, use another storage for the data you neeed and update it there. Modifying DagRun object directly in your hook is a bad idea, this was not what listeners were designed for. Surely you can use screwdriver to hit the nail, but maybe better to use a hammer.

**Reply by potiuk** [reply]
So it's not we do not see your case, we simply strongly advise you to use different solution, because you are shooting yourself in the foot by following something that we strictly discouraged.

**Reply by potiuk** [reply]
> The AIP you keep quoting (AIP 72) isn't even relevant in this discussion. For one, it says nothing about plugins and/or hook.

This AIP is precisely about the fact that we will not let user's provided code to modify and access DB of airlfow. Ever. This is no-go, non-negotiable. The decision has been made. 

It's actually super relevant to your attempt of doing precisely what we want to make impossible in this AIP. And this AIP is (as you likely saw) in-progress and a number of things there need to be clarified still. Possibly worth mentioning that the same limitations (no DB access from user code) apply to hooks and plugins. That's a fair point, and It would be great if you made that comment in the document.

So I (again and again) encourage you to take part in the discussion there, explaining your case and **possibly** seing how your case might be handled differently without accessing SQL Alchemy sessions and Airflow metadata DB, and making sure that it is included there. This is why I attempted to drag your attention to it, because I am aware you might not follow all that. This was my friendly advice where you can look at to understand more and to take part in shaping Airflow future in the way that will allow to handle your case (but without SQLAlchemy and session access).  

And maybe by taking part in that discussion you will understand why we are doing it, and maybe you will also find a way how to approach your case differently in a future-compatible way that you can implement "here and now". This will also make your larg company (and possibly your boss) a bit more happy, because you will no go into a dead-end that you already know is not going to be possible in the future. 

Going this direction and having blind eye on the fact that community discourages it and goes in a different direction than the community chose, might cost your company way more in the long run that investing now in a different approach. 

Treat is as a business advice to you and your company - direct DB access to Airflow DB as part of the user code, is simply bad busness decision.

**Reply by Blizzke** [reply]
> And since we learned that this is causing problems, we continue advising our users to not use sqlalchemy but switch to other means. If REST API is not enough - you probably missed my second suggestion, rather than accessing the DB directly, use another storage for the data you neeed and update it there. Modifying DagRun object directly in your hook is a bad idea, this was not what listeners were designed for. Surely you can use screwdriver to hit the nail, but maybe better to use a hammer.

I did not miss that. 

To start with, again: we are NOT MODIFYING ANYTHING in the database.
We are simply reading some of the lazily loaded relations that happen to not have been loaded before the hook was called. 
So we take that given instance and merge it in another session to do so. 

As for "The REST API": It just seems extremely counter productive to:  receive a `dag_run`-model instance - that has functionality for all the needed attributes and relations - in a hook - that is supported and advertised by the airflow documentation - from *within the actual airflow instance*, and being forced to what? 
Take the ID from it, perform an API call to "yourself" to then get that same data as a json, that you are then forced write code for to fully process *again*, followed by 3 or 4 more calls to obtain the related data, just so you can then start at the same point you were at at the beginning of the callback? Sorry but that simply does not make any sense to me. 

Perhaps you should consider releasing a "airflow models remotely over rest api"-package / aka "airflow data consumer" or something that offers some of the ease of having a models-like interface while hiding the API as much as possible. That could be an easier maintainable (perhaps largely automatically generated) package that makes conversion from "internal" to "external" easier for plugins and other things. That way not everyone has to jump through all those hoops for every thing they want to do.  That might even convince us to switch, but right now I have zero energy to write that stuff myself if the normal way (from my POV) is handy.

But it's clear we will never think the same about this. 
So I guess you can look forward to us shooting ourselves in the foot for the next years.

**Reply by potiuk** [reply]
> We are simply reading some of the lazily loaded relations that happen to not have been loaded before the hook was called.

You are not listening. You will not be able to access SQLAlchemy session. At all. not only for modifying the DB. You are not supposed to interact with DB. Those relations might and will change (and you will have no session to acces them). Don't rely on the fact that they are defined in Airflow code. The way how they are defined are internals.

> As for "The REST API": It just seems extremely counter productive to: receive a dag_run-model instance - that has functionality for all the needed attributes and relations - in a hook - that is supported and advertised by the airflow documentation - from within the actual airflow instance, and being forced to what?

Yes. It's less efficient. This is the trade-off we decided 4 years ago, because thanks to that maintainers are free to modify and treat database as internal detail that users should not rely on. This has been advertised as such since 4 years ago. We explained that. We asked people to tell us what they miss. You are not following what we told you.

> Sorry but that simply does not make any sense to me.

It makes sense to maintainers and community decision was set. You are again arguing here that your needs are more important than the needs of the whole community (and confirmed by vote). It's your opinion vs. deliberate discussion and conscious voting done by multiple people in the community. I think the imbalance is pretty clear.

> Perhaps you should consider releasing a "airflow models remotely over rest api"-package / aka "airflow data consumer" 

This is precisely what is planned in the "irrelevant" as you said AIP-72. Read it closely.

**Reply by Blizzke** [reply]
> You are again arguing here that your needs are more important than the needs of the whole community (and confirmed by vote). 

Please don't twist my words into something that is convenient for you. 
Nowhere has there been said that I think my opinion is more important, it is simply different.

Once more I'm saying that FOR US (and I assume everyone writing plugins against 2.x, or basically anyone that access model data at this moment), this is the easier and cheaper way to go *right now* and that I am not convinced *right now*. Trying to adhere to the new way will force everyone to write API code that gets thrown out again if/when that SDK arrives, which makes investing paid hours in this have zero business value. 

I have also never claimed that I think that this community is going in the wrong direction, nor that I expect that the way we do things right now should be accommodated. I'm simply stating that the way things are done in our company unfortunately do not fit in that vision. Not your fault, nor mine.

The moment an airflow version is released with an SDK to access models and their data we will be the first to migrate to it and refactor everything. I have zero need /passion/excitement doing things this way right now (cfr DB Session), it simply gets me to where I want to be so I can continue with the rest of my work (this is about 1% of it, if that).

> > Perhaps you should consider releasing a "airflow models remotely over rest api"-package / aka "airflow data consumer"
> This is precisely what is planned in the "irrelevant" as you said AIP-72. Read it closely.

Consider me corrected. I've read most of the document diagonally but missed that sdk part I guess, my apologies.

Put a rush on releasing that SDK - so that there is a comparatively equally simple way to consume the API to what there is now plugin-development wise and you'll convince people a lot easier.. That is my advice.

Anyway. This has cost me a lot more of my time already than I was prepared / had allocated to spend on it so I'll just STFU right now. 

We are pretty much living on the edge of the airflow development - forcibly - as we are using the newest features, so I figured that posting about what we encounter could help other people. Guess that was a wrong assumption. So I will stop creating any new discussions/tickets here, as that will save us all a lot of time. Wouldn't want to force my needs on anyone. 
Yes, that was petty, I know.

**Reply by potiuk** [reply]
> Put a rush on releasing that SDK - so that there is a comparatively equally simple way to consume the API to what there is now plugin-development wise and you'll convince people a lot easier.. That is my advice.

We prefer to release things when ready, not rushing it. So - as everything in open-source - we release when ready. But yes if you spend engineering time in understanding and helping in the discussions and maybe even invest engineering efforts in helping to deliver it faster, that would be fantastic way of contributing back to the software your company builds business on for completely free. We release Airflow "for public good" (this is the ASF motto) and many of our users are valuable contributors that give back for the fact that they do not have to implement things on their own. 

That might also help everyone writing plugins to lay their hands on it earlier. And best if those contributions are based on what community discussed and agreed to and voted on, not on personal opinions how things should be (i.e. following "disagree but engage" approach that we practice and preach when we personally disagree with some decisions). This is what we call "community approach".

I hope to see you and your company being part of the efforts and actively contributing, maybe also correcting the course of the community in reaching consensus on how to do things so that you can also use it in the future.

**Reply by kaxil** [reply]
@Blizzke This won't help the discussion about DB access. But for Dataset event extras, that's implemented as part of https://github.com/apache/airflow/issues/37810 (PR: https://github.com/apache/airflow/pull/38650/) and will be released in 2.10.

Docs on main referring to it: http://apache-airflow-docs.s3-website.eu-central-1.amazonaws.com/docs/apache-airflow/stable/authoring-and-scheduling/datasets.html#attaching-extra-information-to-an-emitting-dataset-event


Example:

```python
@task(outlets=[example_s3_dataset])
def write_to_s3(*, outlet_events):
    outlet_events[example_s3_dataset].extra = {"row_count": len(df)}
```
